arcgis-ios-sdk-samples/Content Display Logic/Model/SearchEngine.swift

```diff
@@ -14,165 +14,180 @@
+            return sample.name.lowercased().contains(lowercasedQuery) ||
+                sample.description.lowercased().contains(lowercasedQuery)
+        }
+        // sort matches by relevance
+        matchingSamples.sort { (sample1, sample2) -> Bool in
+            // for convenience, store normalized names for re-use
+            let sample1Name = sample1.name.lowercased()
+            let sample2Name = sample2.name.lowercased()
+            if let sample1Index = sample1Name.range(of: lowercasedQuery)?.lowerBound{
+                if let sample2Index = sample2Name.range(of: lowercasedQuery)?.lowerBound{
+                    // matches are both in the titles
+                    if sample1Index != sample2Index{
+                        // sort by index
+                        return sample1Index < sample2Index
+                    }
+                    // indexes are the same, sort alphabetically
+                    return sample1Name < sample2Name
                 }
+                else{
+                    // only node1 has a title match, sort that first
+                    return true
+                }
+            }
+            else if sample2Name.contains(lowercasedQuery){
+                // only node2 has a title match, sort that first
+                return false
+            }
+            else{
+                // matches are both in the descriptions
+                
+                // for convenience, store normalized descriptions for re-use
+                let sample1Desc = sample1.description.lowercased()
+                let sample2Desc = sample2.description.lowercased()
+                let sample1Index = sample1Desc.range(of: lowercasedQuery)!.lowerBound
+                let sample2Index = sample2Desc.range(of: lowercasedQuery)!.lowerBound
+                if sample1Index != sample2Index{
+                    // sort by index
+                    return sample1Index < sample2Index
+                }
+                // indexes are the same, sort alphabetically
+                return sample1Desc < sample2Desc
             }
         }
+        return matchingSamples
     }
     
-    private func contentOfReadmeFile(directoryPath:String) -> String? {
-        //find the path of the file
-        let path = "\(directoryPath)/README.md"
-        //read the content of the file
-        if let content = try? String(contentsOfFile: path, encoding: String.Encoding.utf8) {
-            return content
+    private func samplesForDisplayNames(_ names:[String]) -> [Sample] {
+        // preserve order
+        return names.compactMap { (name) -> Sample? in
+            return samples.first(where: { (sample) -> Bool in
+                sample.name == name
+            })
         }
-        return nil
     }
     
     //MARK: - Public methods
     
-    func searchForString(_ string:String) -> [String]? {
+    func sortedSamples(matching query:String) -> [Sample] {
```

Comment:
I would suggest returning a `Set<Sample>` and let the caller decide how they want to sort the results, if at all.

Reply:
I think the result sort order is just as important as the results themselves, and should be handled by the search engine. This way the caller won't have to worry about the details, the results are ready to display.

Reply:
But then again should the search engine know how the results are going to be displayed? I don't think so. The caller may decide they want to sort them in descending order or sort by category and then by name. The search engine shouldn't care. It would be one thing if the search engine was ordering the results by ranking, but it is just sorting by name, correct?

Reply:
Nope, I added a non-trivial ranking system internal to the search engine.